### PR TITLE
Megatron-LM integration

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. Integrates Megatron-LM to enable the below optimizations for large scale pre-training/finetuning of BERT (Encoder), GPT (Decoder) or T5 models (Encoder and Decoder). (later will look into extending to VIT and Bi-encoders)
    a. Tensor Parallelism (reduces memory footprint without much additional communication on intra-node ranks)
    b. Pipeline Parallelism (reduces memory footprint and enables large scale training via inter-node parallelization. Also, reduces the bubble of naive PP via `PipeDream-Flush schedule/1F1B schedule` and `Interleaved 1F1B schedule`.  Layers are distributed uniformly across PP stages)
    c. Sequence Parallelism (reduces memory footprint without any additional communication. Only applicable when using TP)
    d. Data Parallelism via Distributed Optimizer (reduces memory footprint by sharding optimizer states and gradients across DP ranks)
    e. Selective Activation Recomputation (reduces the memory footprint of activations significantly via smart activation checkpointing. It doesn't store activations occupying large memory while being fast to recompute thereby achieving great tradeoff between memory and recomputation.)

### To Do
- [ ] Complete coding
- [ ] Test GPT pre-training on instance with 8 GPUs (2 TP X 2 PP X  2 DP) 
- [ ] Write documentation and all the caveats
- [ ] Write tests 
- [ ] Extend to VIT and Bi-Encoders if possible

### CODE REVIEW

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -643,3 +645,142 @@ def load_optimizer(self, accelerator, optimizer, model, input_dir, optimizer_ind
...
+            )
+        if self.sequence_parallelism is None:
+            self.sequence_parallelism = strtobool(os.environ.get(prefix + "SEQUENCE_PARALLELISM", "False")) == 1
+
+        if self.pp_degree > 1:
+            self.DDP_impl = "local"
+        else:
+            self.DDP_impl = "torch"
+
+        self.megtron_lm_default_args = {
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:710]:
```suggestion
        self.megatron_lm_default_args = {
```
Also you use `megatronlm_plugin` with no `_` between `megatron` and `lm`., just flagging this so you stay consistent (I prefer `megatron_lm` personally.

Comment (pacman100) [reply src/accelerate/utils/dataclasses.py]: Done.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -1032,7 +1174,7 @@ def wait_for_everyone(self):
         """
         wait_for_everyone()
 
-    @on_main_process
+    @on_last_process
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1177]: I don't believe all of those should be changed just for MegatronLM.

File: src/accelerate/utils/dataclasses.py [hunk 3]
```diff
@@ -643,3 +645,291 @@ def load_optimizer(self, accelerator, optimizer, model, input_dir, optimizer_ind
...
+        elif "t5" in model.__class__.__name__.lower():
+            model_type_name = "t5"
+            num_layers = model.config.num_layers
+            hidden_size = model.config.d_model
+            num_attention_heads = model.config.num_heads
+            max_position_embeddings = model.config.n_positions
+            orig_vocab_size = model.config.vocab_size
+            pretraining_flag = True
+        else:
+            raise ValueError("Model is not BERT, GPT or T5. Please check the model you are using.")
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:862]: Here the message should be more in the lines of "The Megatron-LM only supports those"

File: src/accelerate/utils/dataclasses.py [hunk 4]
```diff
@@ -643,3 +645,291 @@ def load_optimizer(self, accelerator, optimizer, model, input_dir, optimizer_ind
...
+            "no_wd_decay_cond": self.no_wd_decay_cond,
+            "scale_lr_cond": self.scale_lr_cond,
+            "lr_mult": self.lr_mult,
+            "megatron_dataset_flag": self.megatron_dataset_flag,
+        }
+
+    def set_network_size_args(self, model):
+        # Check if the model is either BERT, GPT or T5 else raise error
+        # set 'num_layers', 'hidden_size', 'num_attention_heads', 'max_position_embeddings'
+        if "bert" in model.__class__.__name__.lower():
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:834]: You should check the `model_type` attribute instead of the class name IMO.

File: src/accelerate/utils/megatron_lm.py [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+    # Initalize and get arguments
+    initialize_megatron(extra_args_provider=extra_args_provider, args_defaults=args_defaults, ignore_unknown_args=True)
+    # Set pytorch JIT layer fusion options and warmup JIT functions.
+    set_jit_fusion_options()
+    args = get_args()
+    args.padded_vocab_size = _vocab_size_with_padding(args.orig_vocab_size, args)
+    if args.model_type_name == "bert" and args.pretraining_flag and args.num_labels == 2:
+        args.bert_binary_head = True
+    # if args.virtual_pipeline_model_parallel_size is not None:
+    #     raise Exception("Virtual pipeline model parallelism is not supported in Accelerate")
```
Comment (sgugger) [inline src/accelerate/utils/megatron_lm.py:157]: To be cleaned up?

**Review by sgugger — COMMENTED** [review]
Looks good overall, thanks a lot for diving into this! Would be great to have some documentation to go along with this new integration.

File: src/accelerate/accelerator.py [hunk 6]
```diff
@@ -71,6 +73,21 @@
...
+        MegatronLMDummyDataLoader,
+        MegatronLMDummyScheduler,
+        MegatronLMOptimizerWrapper,
+        MegatronLMSchedulerWrapper,
+    )
+    from .utils import initialize as megatron_lm_initialize
+    from .utils import prepare_data_loader as megatron_lm_prepare_data_loader
+    from .utils import prepare_model as megatron_lm_prepare_model
+    from .utils import prepare_optimizer as megatron_lm_prepare_optimizer
+    from .utils import prepare_scheduler as megatron_lm_prepare_scheduler
```
Comment (sgugger) [inline src/accelerate/accelerator.py:88]: The aliases should be done in the `.utils` module: if we have another integration like this, we won't be able to have both names.

File: src/accelerate/utils/__init__.py [hunk 7]
```diff
@@ -94,6 +96,23 @@
...
+    MegatronLMDummyScheduler,
+    MegatronLMOptimizerWrapper,
+    MegatronLMSchedulerWrapper,
+    T5TrainStep,
+    avg_losses_across_data_parallel_group,
+    initialize,
+    prepare_data_loader,
+    prepare_model,
+    prepare_optimizer,
+    prepare_scheduler,
```
Comment (sgugger) [inline src/accelerate/utils/__init__.py:114]: AS said above, those should probably have the `megatron_lm` prefix here.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this! LGTM, just have a couple of comments re-names.

File: src/accelerate/accelerator.py (hunk 6 above)
Comment (pacman100) [reply src/accelerate/accelerator.py]: Done.

File: src/accelerate/utils/__init__.py (hunk 7 above)
Comment (pacman100) [reply src/accelerate/utils/__init__.py]: Done.

File: docs/source/usage_guides/megatron_lm.mdx [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+
+a. **Tensor Parallelism (TP)**: Reduces memory footprint without much additional communication on intra-node ranks
+
+b. **Pipeline Parallelism (PP)**: Reduces memory footprint and enables large scale training via inter-node parallelization. 
+Reduces the bubble of naive PP via PipeDream-Flush schedule/1F1B schedule and Interleaved 1F1B schedule. 
+Layers are distributed uniformly across PP stages.
+
+c. **Sequence Parallelism (SP)**: Reduces memory footprint without any additional communication. Only applicable when using TP.
+
+d. **Data Parallelism (DP)** via Distributed Optimizer: Reduces memory footprint by sharding optimizer states and gradients across DP ranks.
```
Comment (sgugger) [inline docs/source/usage_guides/megatron_lm.mdx:34]: Can be for another PR, but would be nice to explain the difference between those kinds of parallelism and what is parallelized exactly.

**Review by sgugger — APPROVED** [review]
LGTM!

File: src/accelerate/utils/operations.py [hunk 9]
```diff
@@ -224,6 +224,7 @@ def gather(tensor):
         DistributedType.DEEPSPEED,
         DistributedType.MULTI_GPU,
         DistributedType.FSDP,
+        DistributedType.MEGATRON_LM,
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/utils/operations.py:227]: PR #758 to resolve this

File: docs/source/usage_guides/megatron_lm.mdx (hunk 8 above)
Comment (pacman100) [reply docs/source/usage_guides/megatron_lm.mdx:34]: PR #759 to address this.
